ActivityPub: add compatibility functions (targeting Photon)

jeherve opened the pull request:
> **Warning**
> Do not merge until #30050 has been merged.

## Proposed changes:

The ActivityPub plugin returns an object with information about each post when you make a ActivityPub request. This includes an image for the post, fetched using core's `wp_get_attachment_image_src`.

Jetpack's Image CDN changes the URL of the images returned by that function. In some scenarios, that may not be desirable (Some Fediverse instances may not always be able to fetch images from the Jetpack CDN, for example).

In this PR, we add a new compatibility layer for the ActivityPub plugin, where we allow disabling Jetpack's Image CDN processing when fetching an image for ActivityPub. This is only enabled if you use a filter:

```php
add_filter( 'jetpack_activitypub_post_disable_cdn', '__return_true' );
```

**Discussions**

- https://github.com/pfefferle/wordpress-activitypub/pull/307
- https://github.com/pfefferle/wordpress-activitypub/pull/309

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* Internal reference: p1682284516722569-slack-C04TJ8P900J

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

* Go to Jetpack > Settings > Performance and ensure that the Site Accelerator feature is enabled.
* Install and activate the current development version of the ActivityPub plugin, from https://github.com/automattic/wordpress-activitypub/
* Publish a post that includes an image.
* Make a GET request for that post, with the `Accept: application/activity+json` header.
* You should see the image using a `i0.wp.com` prefix in the response
* Now add the above snippet in a functionality plugin, or in your theme's `functions.php` file.
Make the same request again; the `i0.wp.com` prefix should be gone.

jeherve committed:
[not verified] ActivityPub: add compatibility functions

The ActivityPub plugin returns an object with information about each post when you make a ActivityPub request. This includes an image for the post, fetched using core's wp_get_attachment_image_src.

Jetpack's Image CDN changes the URL of the images returned by that function. In some scenarios, that may not be desirable (Some Fediverse instances may not always be able to fetch images from the Jetpack CDN, for example).

In this commit, we add a new compatibility layer for the ActivityPub plugin, where we allow disabling Jetpack's Image CDN processing when fetching an image for ActivityPub. This is only enabled if you use a filter.

Discussions:

- https://github.com/pfefferle/wordpress-activitypub/pull/307
- https://github.com/pfefferle/wordpress-activitypub/pull/309

Internal reference: p1682284516722569-slack-C04TJ8P900J

jeherve committed:
Add @version numbers

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/compat-activitypub

anomiex commented on projects/plugins/jetpack/3rd-party/3rd-party.php:
```diff
@@ -17,6 +17,11 @@
  */
 function load_3rd_party_compat_filters() {
+	// ActivityPub.
+	if ( class_exists( '\Activitypub\Model\Post' ) ) {
```
Nitpick, feel free to ignore
```suggestion
	if ( class_exists( \Activitypub\Model\Post::class ) ) {
```

anomiex reviewed (commented):
Looks ok, but should it somehow address Boost too?

jeherve committed:
Update projects/plugins/jetpack/3rd-party/3rd-party.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Move ActivityPub compat logic to the package

If it's in the package, it will be triggered regardless of the plugin using the package.

See https://github.com/Automattic/jetpack/pull/30298#pullrequestreview-1414949400

jeherve committed:
Add changelog

jeherve commented:
> Looks ok, but should it somehow address Boost too?

That's a good point, I forgot about Boost shipping with the CDN now. Maybe this compatibility layer should live within the package so it's available to anyone using it. I've just pushed b5f7ce15262dd7015d547931c29ac52f00d86b9a to do that.

anomiex reviewed (dismissed):
Looks sensible. Haven't tested.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/compat-activitypub

jeherve committed:
Bump version

zinigor reviewed (approved):
It's safe to merge this now.
